Add asking effective questions page (#432)
gamemanual0/gm0: source/docs/being-a-team/how-to-ask-effective-questions.rst

.. include:: <isonum.txt>

How to Ask Effective Questions
==============================

Asking questions effectively is one of the most important skills to develop, especially for your life outside of robotics.
Being able to learn from your questions properly is crucial and as such, you should be able to ask good questions, help those who want to help you, and make the most of what you are taught.

> Research First

Reviewer: I would agree with previous comments that this should be it's own page.

--------------

Before you ask a question, you should be doing your own research to find the answer.
This will take more time, it will take more effort, but it is a crucial skill to learn to find solutions on your own, and learn how to research effectively to find solutions to your problems.
In the end, this research will pay off in the long run, helping you to learn about topics relevant to your question, and often helping you to understand the roots of the problems you encounter.

Here are some steps that you could take to research your problem before asking someone a question about your current problem.

> - **Google** - Simply search your problem.

Reviewer: ```suggestion
-**Checking Official Documentation** - If documentation or instructions are available, make sure to carefully examine the relevant sections of them . Your problem may already be addressed, or may be caused by you misreading something.

- **Google** - Simply search your problem.
```
 I'm not sure whether I'm putting this in exactly the right order but do think mentioning official docs is important.

  You will be surprised by how often this can quickly solve your problem although you may not have expected it to.

>   Try word the question in many different ways to find similar questions that may have already been asked.

Reviewer: Maybe mention stuff like searching using keywords and specific error messages here?

.. tip::

>     Add "filetype:pdf" to the beginning of your search to filter the search results into only .pdf documents.

Reviewer: I would generally recommend against this for programming issues

Author: Clarified this for use with documents and textbooks.

    This can help you when looking for past documents or sections of textbooks.

> - **Youtube** - Just like google, there are many great resources on Youtube which can help with very specific, obscure problems.

Reviewer: ```suggestion
- **YouTube** - There are many great resources on YouTube which can help with even obscure problems.
```

> - **Reddit and Discord** - If you can't find your own question online, the FTC Reddit and Discord will be the other most likely place to have a past question similar to your own.

Reviewer: ```suggestion
- **Reddit and Discord** - The FTC [subreddit](https://www.reddit.com/r/ftc/) and [Discord server](https://discord.gg/first-tech-challenge) can also be useful places to find previous answers or to ask your own.
```

>   Try searching through the past questions with words related to your question and see if a similar question has been asiked before.

Reviewer: ```suggestion
  Try searching through the past questions with keywords relating to your own.
```

>   You may find information in this discussion which can help you to solve your own problem.

Reviewer: ```suggestion
```

> - **Past Examples** - Try find past implementations of whatever problem you are trying to solve.

Reviewer: ```suggestion
- **Past Examples** - For programming questions, it can sometimes be helpful to find past implementations of whatever problem you are trying to solve through [GitHub Code Search](https://github.com/search?type=code&auto_enroll=true).
```

Reviewer: This could be used for mech stuff too, but not really sure where you'd find things in a similar way to code search.

Author: Extracted out into a tip to make it more clear.

>   This step may take more time to find the example, and then decipher how someone else figured out the solution, but this process can be super valuable.

Reviewer: ```suggestion
Even if you don't find the solution, examining other codebases can teach you a lot about general strategies and functions you may not have seen before.
```

>   For programming, digging through another person's repo can be very insightful, and likewise, for mechanical design, taking a look at a teams technical binder or CAD release.

Reviewer: ```suggestion
  For programming, digging through another person's repository can be very insightful. 
 Likewise, for mechanical design, taking a look at a teams technical binder or CAD release can give you a lot of new ideas.
```

Be Specific
-----------

For others to be able to provide a good solution to your problem, they need the specifics.
This means doing the following:

1. **Break down the problem.** You should try to pinpoint the problem as much as possible.
For example, when encountering robot crashes, try to collect as much data as possible around the times when your robot crashes.

> You may not understand this information yourself, but it can be a huge help to others trying to diagnose your issue.

Reviewer: ```suggestion
You may not understand this information yourself, but it may allow others to recognize a pattern and solve your issue.
```

2. **Provide some context.** It is often helpful to provide context for example,  what skills you are new to and hence, could be the root of the issues.

> Furthermore, information about your resources can help others to determine solutions which are feasible and effective within your own constraints.

Reviewer: ```suggestion
Furthermore, information about your resources can help others to determine solutions which are feasible and effective within your own constraints.
For example, you might be competing within a week and looking for a quick fix; make sure to mention this in your question if so.

It can also be useful to explain why you're asking. It can be easy to get stuck on a specific solution to a problem when it would be better solved in an entirely different way; this is known as the [XY problem.](https://xyproblem.info/)
```
Here's a start on a mention of XY problem, not sure whether this is as clear as it could be or in the right spot though.

Author: Added to step 1, break down the problem.

3. **Share your progress.** Explain the steps you have already taken to address the problem.
This may include specific troubleshooting methods, or examples of what you have been changing to try solve the issue.
This helps others to gain a better understanding of the problem and avoid suggesting solutions that you have already tried.

4. **Don't just use words.** You should always try to provide some additional media related to the problem.
This may be sections of code, CAD files, images or videos.

> All of these provide others with a much better understanding of your problem, and often provide significantly useful information, which could be seemingly unrelated.

Reviewer: ```suggestion
All of these provide others with a much better understanding of your problem.
For example, if you're experiencing an issue tuning Roadrunner, be sure to include your tuning values and localizer configuration. 
```

Understand the Solution
-----------------------

The final step for asking a question is understanding the solution.

> This is your own responsibility and no matter how good a teacher is, if you do not ask clarifying questions, you limit your own learning.

Reviewer: ```suggestion
This is your own responsibility. 
No matter how good a teacher is, if you do not ask clarifying questions, you limit your own learning.
```

To really understand the solution, you should do the following.

- **Understand why the problem is solved.** How was your problem fixed? Why does this fix work?
- **Figure out why the problem occurred.** What was the root cause of the problem? How can you stop it happening again?

> - **Ask for an outline of someone elses problem solving process.** How did someone else solve the problem? How can you use this process to solve more of your own problems in the future?

Reviewer: ```suggestion
- **Ask for an outline of someone else's problem solving process.** How did someone else solve the problem? How can you use this process to solve more of your own problems in the future?
```

.. include:: <isonum.txt>

How to Ask Effective Questions
==============================

> Asking questions effectively is one of the most important skills to develop, especially for your life outside of robotics.
> Being able to learn from your questions properly is crucial and as such, you should be able to ask good questions, help those who want to help you, and make the most of what you are taught.

Reviewer 2: ```suggestion
Asking questions effectively is one of the most important skills to develop, especially for your life outside of robotics. Being able to learn from your questions properly is crucial and as such, you should be able to ask good questions, help those who want to help you, and make the most of what you are taught.
```

We write paragraphs as one line, because that makes git diffs nicer. This is an example, it should be done for every paragraph.

In the end, this research will pay off in the long run, helping you to learn about topics relevant to your question, and often helping you to understand the roots of the problems you encounter.

Here are some steps that you could take to research your problem:

- **Checking Official Documentation** - If documentation or instructions are available, make sure to carefully examine the relevant sections of them.
  Your problem may already be addressed, or may be caused by you misreading something.

- **Google** - Simply search your problem.
  You will be surprised by how often this solves the issue.
  Try word the question in many different ways to find similar questions that may have already been asked.
  Focus on including keywords about your problem and if it is about programming, include relevant error messages and codes.

.. tip::
    When looking for past documents or sections of textbooks, add "filetype:pdf" to the beginning of your search to filter the search results into only .pdf documents.

- **YouTube** - There are many great resources on YouTube which can help with even obscure problems.

> - **Reddit and Discord** - The `FTC Subreddit <https://www.reddit.com/r/ftc/>`_ and `Discord Server <https://discord.gg/first-tech-challenge>`_ can also be useful places to find previous answers or to ask your own.

Reviewer 2: We link to and correspondingly warn about the FTC Discord elsewhere (https://gm0.org/en/latest/docs/useful-resources.html), which I think is important. it might be useful to either take this warning or to link to that reference instead of the server directly.

Author: Fixed in most recent commit 
[https://github.com/gamemanual0/gm0/pull/432/commits/c2f3ae902bb6f276fcf27a23e403ce33948c3e9b](url)

  Try searching through the past questions with keywords relating to your own.

- **Past Examples** - Try find past implementations of whatever problem you are trying to solve.

>   This step may take more time to find the example, and then decipher how someone else figured out the solution, but this process can be super valuable.

Reviewer 2: ```suggestion
  This step may take more time, as it requires finding the example, and then deciphering how someone else figured out the solution, but this process can be super valuable.
```
Slight change the sentence to make it flow better.

  For programming, digging through another person's repository can be very insightful.
  Likewise, for mechanical design, taking a look at a teams technical binder or CAD release.

.. tip::

>     For programming questions, it can sometimes be helpful to find past implementations through `GitHub Code Search <https://github.com/search?type=code&auto_enroll=true>`_.

Reviewer 2: ```suggestion
    For programming questions, it can sometimes be helpful to find past implementations through `GitHub Code Search <https://github.com/search?type=code>`_. Note this currently requires a GitHub account.
```

    Even if you don't find the solution, examining other codebases can teach you a lot about general strategies and functions you may not have seen before.

Be Specific
-----------

For others to be able to provide a good solution to your problem, they need the specifics.
This means doing the following:

1. **Break down the problem.**

> You should try to pinpoint the problem as much as possible.
> For example, when encountering robot crashes, try to collect as much data as possible around the times when your robot crashes.
> You may not understand this information yourself, but it may allow others to solve your issue.
> It can also be useful to explain why you're asking.
> It can be easy to get stuck on a specific solution to a problem when it would be better solved in an entirely different way; this is known as the `XY Problem <https://xyproblem.info/>`_.

Reviewer 2: ```suggestion
   You should try to pinpoint the problem as much as possible. For example, when encountering robot crashes, try to collect as much data as possible around the times when your robot crashes. You may not understand this information yourself, but it may allow others to solve your issue. It can also be useful to explain why you're asking. It can be easy to get stuck on a specific solution to a problem when it would be better solved in an entirely different way; this is known as the `XY Problem <https://xyproblem.info/>`_.
```
The paragraphs here should also be formatted as one line, but should also be indented to be properly contained within the list.

4. **Don't just use words.**

You should always try to provide some additional media related to the problem.
This may be relevant sections of code, CAD files, images or videos.
All of these provide others with a much better understanding of your problem.

Understand the Solution
-----------------------

The final step for asking a question is understanding the solution.
This is your own responsibility.
No matter how good a teacher is, if you do not ask clarifying questions, you limit your own learning.
To really understand the solution, you should do the following.

- **Understand why the problem is solved.** How was your problem fixed? Why does this fix work?
- **Figure out why the problem occurred.** What was the root cause of the problem? How can you stop it happening again?
- **Ask for an outline of someone else's problem solving process.** How did someone else solve the problem? How can you use this process to solve more of your own problems in the future?
- **Find gaps in your knowledge.** What information were you missing that led to the problem? How can you develop this area of your knowledge?

.. tip::

>     If the answer is not immediately obvious to one of the steps above, ask the person who helped you and have them provide some more insight.

Reviewer 2: ```suggestion
    If the answer to one of these questions is not apparent, ask the person who helped you and have them provide some more insight.
```
Slight wording tweak.
